### PR TITLE
Add a structured verification results class in the C++ API

### PR DESCRIPTION
**Pull request by NlightNFotis**
## One-sentence Description

This PR is adding the capacity to the C++ API to provide the end-results of a verification engine run to a client in a structured format.

## Review Guidance

* The gist of this change is the addition of the `libcprover-cpp/verification_result.h` file, along with the associated infrastructure in `all_properties_verifier_with_trace_storaget` class (moving the traces from the verifier at the end of the verification engine run).
* This was initially added in `src/goto-checker/`, but offline discussion with @peterschrammel identified that as a potentially disruptive change. It was then decided to move this closer to more experimental code in the C++ api, rather than keep it close to the solvers.
* To support testing of this new capacity, a new helper function has been added in `testing-utils/run_verification_engine.h` - it clones the functionality of the C++ API `verify_model()` function with the aim to quickly do a verification engine run in a convenient package and produce back the verification results that we expect from the engine.

## Checklist

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by martin-cs — COMMENTED** [review]
I hate being That Guy but...

How does this fit with the `goto_verifier` / `properties` interface?

https://github.com/diffblue/cbmc/blob/develop/src/goto-checker/goto_verifier.h
https://github.com/diffblue/cbmc/blob/develop/src/goto-checker/properties.h

I know @peterschrammel did some work on setting this up as the future of "unified verification results".  I did some work on getting the abstract interpreter to fit with it:

https://github.com/diffblue/cbmc/blob/develop/src/goto-analyzer/static_verifier.h

but it needs more work.

**Comment by NlightNFotis** [comment]
Hi Martin, @martin-cs 

The motivation for this is to have an interface from which the API clients will be able to get the results of a run from the verification engine and inspect them.

~With regards to `goto_verifiert` the long term plan is to refactor the `operator()` to be able to return these structured results (which is really a triple `<resultt, propertiest, goto_trace_storaget>` - we already have this information and accessors for these in the various subclasses) but this is slightly more involved and I want to build the functionality for the API first, both as a test-bed and as a capability for these APIs, in a non-invasive way.~

~It's worth mentioning that for now this only exists in one configuration, the one that is used by the test. The next PR will add support for this in the C++ API (under `src/libcprover-cpp`), and then it's going to be exported to the Rust API.~

---

I need to sort out CI here first and then I will clean up the commit history and also provide some review guidance in the description above as to what changes were made and why.

---

**Update**: I have crossed out the part of the text that contains some invalid assumptions of mine after some discussion with Peter. The code will need some rework, so as it stands right now it doesn't reflect the final design.

**Comment by NlightNFotis** [comment]
Update:

I had a discussion with @peterschrammel offline about parts of the design and it contains some defects that I was oblivious too, which renders my previous comment invalid.

This is a change only for `all_properties_goto_verifier_with_trace_storaget` and can't affect objects from the `goto_verifiert` hierarchy that don't feature trace storage, so expanding the API of the class to support this would lead to an API design that's confusing.

I will rework the API to move the `produce_results()` method to the C++ directly rather than augment the interface of that specific verifier.

**Comment by TGWDB** [comment]
Fixes #7499

File: src/libcprover-cpp/api.cpp
Comment (thomasspriggs) [inline src/libcprover-cpp/api.cpp]: I think `return;` on failure was already the correct behaviour, as errors are supposed to be reported via the `log` variable.

File: src/libcprover-cpp/api.h
Comment (thomasspriggs) [inline src/libcprover-cpp/api.h]: :pick: This should have three forward slashes `///` for doxygen purposes.

File: unit/testing-utils/message.h
Comment (thomasspriggs) [inline unit/testing-utils/message.h]: :pick: The justification for moving this belongs in the commit message body, not a comment.
:pick: You haven't actually shown that there is a more general use for this class, given that you haven't added new uses of it within this PR.

File: unit/testing-utils/message.h
Comment (thomasspriggs) [inline unit/testing-utils/message.h]: :no_entry_sign: This `get_ui` member function hides the member function of the same name in the parent class, but is presumably supposed to `override` it? Same comment applies to the `get_json_stream` function and the `test_ui_message_handlert_plain::get_ui` member function of your new class.

File: src/libcprover-cpp/verification_result.h
Comment (thomasspriggs) [inline src/libcprover-cpp/verification_result.h]: :no_entry_sign: API headers must not include internal headers, because this exposes implementation details as part of the public API.

File: src/goto-checker/all_properties_verifier_with_trace_storage.h
Comment (thomasspriggs) [inline src/goto-checker/all_properties_verifier_with_trace_storage.h]: :pick: This method has "move" in its name, but it doesn't move or have any `&&` types, it just provides mutable access.

File: unit/testing-utils/run_verification_engine.cpp
Comment (thomasspriggs) [inline unit/testing-utils/run_verification_engine.cpp]: :pick: The coding standard says that abbreviations should be avoided.

File: src/goto-checker/goto_trace_storage.h
Comment (thomasspriggs) [inline src/goto-checker/goto_trace_storage.h]: I am not sure this is a good idea as it could lead to accidental copy construction of `goto_trace_storaget` which can be an extremely large data structure. Think multiple gigabytes...

File: src/goto-checker/all_properties_verifier_with_trace_storage.h
Comment (thomasspriggs) [inline src/goto-checker/all_properties_verifier_with_trace_storage.h]: :pick: The rest of the codebase uses `\brief`. Using `@brief` here seems kinda incongruous. The same comment applies to other usages of `@brief` in this PR.

File: src/libcprover-cpp/api.cpp
Comment (thomasspriggs) [inline src/libcprover-cpp/api.cpp]: :pick:  It might be worth adding a comment above this that `all_properties_verifier_with_trace_storaget` doesn't store traces without this.
:thinking: Ultimately, `all_properties_verifier_with_trace_storaget` is a pretty small class which doesn't fit the needs of the API very well. So we are most likely better off writing our own version of it.

File: unit/testing-utils/message.h
Comment (thomasspriggs) [inline unit/testing-utils/message.h]: I think that adding a new class to the hierarchy actually makes the tests unnecessarily difficult to follow when the same could be achieved with just `ui_message_handlert{null_message_handler}`

**Review by thomasspriggs — CHANGES_REQUESTED** [review]
Some initial comments rather than a complete review, given that this likely needs substantial work to deal with this comment -

- :no_entry_sign: API headers must not include internal headers, because this exposes implementation details as part of the public API.

File: unit/testing-utils/run_verification_engine.cpp
Comment (peterschrammel) [inline unit/testing-utils/run_verification_engine.cpp]: Why doesn't use this the API function to test?

File: src/libcprover-cpp/api.cpp
Comment (peterschrammel) [inline src/libcprover-cpp/api.cpp:179]: 💡 I'd prefer documentation in the .h file because this is where the interface is defined.

File: src/libcprover-cpp/verification_result.h
Comment (NlightNFotis) [reply src/libcprover-cpp/verification_result.h]: Move to .cpp with `pimpl`.

File: unit/libcprover-cpp/CMakeLists.txt
Comment (NlightNFotis) [inline unit/libcprover-cpp/CMakeLists.txt]: See if it is to be removed - we don't want to expose symbols from the `libcproverunit` binary.

File: src/goto-checker/goto_trace_storage.h
Comment (NlightNFotis) [reply src/goto-checker/goto_trace_storage.h]: See if I can change the class that manages this to be holding a unique pointer which can then be moved.

File: src/goto-checker/all_properties_verifier_with_trace_storage.h
Comment (NlightNFotis) [reply src/goto-checker/all_properties_verifier_with_trace_storage.h]: This is no longer relevant - I have removed the method added, given that the traces are now outside of the scope of this PR.

File: unit/testing-utils/message.h
Comment (NlightNFotis) [reply unit/testing-utils/message.h]: All of the changes here have been reverted because of the simplification you proposed below:

```
ui_message_handlert{null_message_handler}
```

File: unit/testing-utils/message.h
Comment (NlightNFotis) [reply unit/testing-utils/message.h]: Done.

File: src/libcprover-cpp/verification_result.h
Comment (thomasspriggs) [inline src/libcprover-cpp/verification_result.h:28]: :pick: Seems like quite a generic symbol to introduce into the public global namespace.

File: src/libcprover-cpp/api.cpp
Comment (thomasspriggs) [reply src/libcprover-cpp/api.cpp:179]: This is an actual TODO comment rather than documentation. It is explaining the need for clean-up rather than documenting what the function is supposed to be for. Therefore I would argue that this information belongs in the `.cpp` rather the header. ~~There is a separate argument that there should be a doxygen block in the header which  deprecates this function, explains the difference from the `verify_model` function and that the `verify_model` function should be used in preference.~~ The is already a doxygen block in the header.

File: src/libcprover-cpp/api.cpp
Comment (thomasspriggs) [reply src/libcprover-cpp/api.cpp]: :raised_eyebrow: You gave this a thumbs up but haven't actioned it?

Comment (thomasspriggs) [reply src/libcprover-cpp/api.cpp]: Using `INVARIANT` here is actually a really bad thing in terms of an API. This is because if the INVARIANT fails it will either abort the whole process which is undesirable if the consumer on the API can recover or it will throw an exception which may not be supported by the ABI being used to interface with the API.

File: unit/testing-utils/message.h
Comment (thomasspriggs) [reply unit/testing-utils/message.h]: Neat.

File: src/libcprover-cpp/verification_result.h
Comment (thomasspriggs) [reply src/libcprover-cpp/verification_result.h]: Sorted.

File: src/goto-checker/all_properties_verifier_with_trace_storage.h
Comment (thomasspriggs) [inline src/goto-checker/all_properties_verifier_with_trace_storage.h]: This reformatting appears to be the result of adding and removing things in the process of work on this PR. If you removed all the changes to this file, then you would not need as many code-owners to approve...

File: src/goto-checker/all_properties_verifier_with_trace_storage.h
Comment (thomasspriggs) [reply src/goto-checker/all_properties_verifier_with_trace_storage.h]: This method is gone for the moment.

File: src/libcprover-cpp/api.cpp
Comment (NlightNFotis) [reply src/libcprover-cpp/api.cpp]: This is no longer relevant, as marked `outdated` by the GitHub UI itself.

This change has been rolled back in b9b6cc4aa151b09f3a4407c769df4201cb4ed888 - https://github.com/diffblue/cbmc/commit/b9b6cc4aa151b09f3a4407c769df4201cb4ed888#diff-122d9d2e65523b48e18239983a025b547f9509d43e638fea6acc71a926daed69R155

File: unit/testing-utils/module_dependencies.txt
Comment (thomasspriggs) [inline unit/testing-utils/module_dependencies.txt]: I think the functionality related to testing `src/libcprover-cpp` should live in `unit/libcprover-cpp` not `unit/testing-utils`. This is particularly the case with the API, as it is desirable to be able to unit test the API without linking the API unit test binary to the rest of the code base on a `#include` level.

File: unit/libcprover-cpp/verification_resultt.cpp
Comment (thomasspriggs) [inline unit/libcprover-cpp/verification_resultt.cpp]: :no_entry_sign: This is introducing a transitory include of `goto-programs/goto_model.h`  which is not part of the API. It must be possible to test the API without using the rest of the codebase.

File: src/libcprover-cpp/verification_result.cpp
Comment (thomasspriggs) [inline src/libcprover-cpp/verification_result.cpp:137]: :bulb: I suggest making the error case the default case. This will avoid compile errors about reaching the end of non-void functions.

**Review by thomasspriggs — CHANGES_REQUESTED** [review]
I think we should fix the inclusion of private headers in the API unit test before merging as a continuation of the fix to avoid including private headers in the public API headers.

File: src/libcprover-cpp/api.cpp
Comment (thomasspriggs) [reply src/libcprover-cpp/api.cpp]: This still needs fixing in the `produce_results` function. This comment was in the `verify_model` function but the code is duplicated, so the fix needs doing in two places.

File: unit/testing-utils/run_verification_engine.cpp
Comment (thomasspriggs) [reply unit/testing-utils/run_verification_engine.cpp]: This seems to have been fixed.

File: src/libcprover-cpp/api.cpp
Comment (NlightNFotis) [reply src/libcprover-cpp/api.cpp]: Good catch - my bad. I did a pattern replacement to avoid this issue, but for a strange kind of fashion this was missed.

File: src/libcprover-cpp/verification_result.cpp
Comment (NlightNFotis) [reply src/libcprover-cpp/verification_result.cpp:137]: I prefer to keep the `UNREACHABLE` - this is to communicate intent about this being an exhaustive list.

It also guards against a new enum member being added but not being matched against properly, as it will fail fast and loud.

Comment (thomasspriggs) [reply src/libcprover-cpp/verification_result.cpp:137]: As I'd previously pointed out in one of my other comments on this PR, the standard `INVARIANT` / `UNREACHABLE` macros don't work in a useful fashion in this context. Therefore an alternative means of reporting the error is required. I suggest using a comment if you feel the need to explain the the default case should be unreachable and is therefore an error.

File: unit/libcprover-cpp/verification_resultt.cpp
Comment (thomasspriggs) [inline unit/libcprover-cpp/verification_resultt.cpp]: If this PR isn't using the `run_verification_engine` function anymore then it should not add it.

File: src/libcprover-cpp/verification_result.cpp
Comment (thomasspriggs) [inline src/libcprover-cpp/verification_result.cpp:67]: :pick: This should be `= default`.

File: src/libcprover-cpp/verification_result.h
Comment (thomasspriggs) [inline src/libcprover-cpp/verification_result.h]: :exclamation: The type signature of this operation makes no sense. `&&` states that the operator is moving the data from `other` which would mutate `other`. However `const` states that the operator does not mutate `other`. I think the appropriate way to resolve this is just to remove this overload entirely.

File: src/libcprover-cpp/verification_result.cpp
Comment (thomasspriggs) [inline src/libcprover-cpp/verification_result.cpp:18]: :thinking: I think you can just make this class a struct, delete all the member functions and just use the fields directly. It would be far more straight-forward.

File: src/libcprover-cpp/verification_result.cpp
Comment (thomasspriggs) [inline src/libcprover-cpp/verification_result.cpp]: :no_entry_sign: The type `dstringt` should not be used here. The type of the key of the properties collection is an `irep_idt`. The typedef may define `irep_idt` to be either a `dstringt` or `std::string` depending on the build options. So if you use `dstringt` here then you will break the `std::string` build.

This comment also applies to the forward declaration of type `propertiest` in `verification_result.h`.

File: src/libcprover-cpp/verification_result.h
Comment (NlightNFotis) [reply src/libcprover-cpp/verification_result.h]: Good catch - this was likely a copy-paste fail.

I think this shouldn't be removed - the intent here is to implement copy assignment.

File: src/libcprover-cpp/verification_result.cpp
Comment (NlightNFotis) [reply src/libcprover-cpp/verification_result.cpp:18]: I'd rather keep the API here as it stands - to be a bit more defensive in case tighter access control (say, some form of translation of the types, or validation) is required in the future (which can be implemented transparently through the property methods).

If that ends up not being needed, the accessor methods should be able to be removed with little effort.

Comment (thomasspriggs) [reply src/libcprover-cpp/verification_result.cpp:18]: I disagree with your reasoning. "impl" means implementation detail; it is in the name. It shouldn't be intended to be part of the exposed API. Therefore the using a class to add accessibility restrictions to something which is already private by virtue of being declared in the `.cpp` only is overkill.

This is not a blocking comment, just an observation. So it can be left as-is for the moment if you prefer.

File: src/libcprover-cpp/verification_result.h
Comment (thomasspriggs) [reply src/libcprover-cpp/verification_result.h]: :+1:

**Review by thomasspriggs — APPROVED** [review]

**Pull request merged**
